File: FishTrackerLambda/src/FishTrackerLambda/Functional/Function.cs
using System;
using System.Collections.Generic;

namespace FishTrackerLambda.Functional
{
	public static class Function
	{
        public static Task<HttpWrapper<object>> ValidateInput(Func<IResult?> result)
        {
            var res = result();
            return Task.FromResult(res == null ? HttpWrapper<object>.Ok(new object()) : HttpWrapper<object>.FromResult(res));
        }

        public static Task<HttpWrapper<object>> ValidateInput(this Task<HttpWrapper<object>> record, Func<IResult?> result)
        {
            return ValidateInput(result);
        }

        public static Task<HttpWrapper<T>> Init<T>(T value)
        {
            return Task.FromResult(new HttpWrapper<T>(value));
        }

        public static Task<HttpWrapper<T>> InitAsync<T>(Task<HttpWrapper<T>> record)
        {
            return record;
        }

        public static Task<HttpWrapper<R>> Init<T,R>(this Task<HttpWrapper<T>> record, R value)
        {
            return Task.FromResult(new HttpWrapper<R>(value));
        }

        public static async Task<HttpWrapper<R>> MapAsync<T,R>(this Task<HttpWrapper<T>> record, Func<T, Task<HttpWrapper<R>>> mapper)
		{
			var waitedRec = await record;

			if (waitedRec.Continue == false)
			{
				var clone = waitedRec.CloneFailed<R>();
                return clone;
			}

            T? value = waitedRec.Value;

            return value != null ? await mapper(value) : throw new Exception("Mapping null value");
		}

        public static async Task<HttpWrapper<IEnumerable<R>>> MapEachAsync<T, R>(this Task<HttpWrapper<IEnumerable<T>>> record, Func<T, Task<HttpWrapper<R>>> mapper)
        {
            var waitedRec = await record;

            if (waitedRec.Continue == false)
            {
                var clone = waitedRec.CloneFailed<IEnumerable<R>>();
                return clone;
            }

            IEnumerable<T>? values = waitedRec.Value;
            if (values == null)
                throw new Exception("Mapping null value");

            var d = values.Select(v => mapper(v));
            var all = await Task.WhenAll(d);

            var failed = all.Where(f => f.Continue == false).Select(s => s.Result);
            IEnumerable<R> success = all.Where(f => f.Continue == true && f.Value != null).Select(s => {
                return s.Value ?? throw new Exception("value should npt be null");
            });

            return failed.Any() ? HttpWrapper<IEnumerable<R>>.FromResult(failed.First()) : HttpWrapper<IEnumerable<R>>.Ok(success);
        }

        public static async Task<HttpWrapper<R>> Map<T, R>(this Task<HttpWrapper<T>> record, Func<T, R> mapper)
        {
            var waitedRec = await record;

            if (waitedRec.Continue == false)
            {
                var clone = waitedRec.CloneFailed<R>();
                return clone;
            }

            T? value = waitedRec.Value;

            return value != null ? new HttpWrapper<R>(mapper(value)) : throw new Exception("Mapping null value");
        }
    }
}

File: FishTrackerLambdaNode/src/Functional/Function.ts
import { HttpWrapper } from './HttpWrapper';
import { Results } from '../Http/Result';

export class Function {
    static async ValidateInput(result: () => Results | null): Promise<HttpWrapper<object>> {
        const res = result();
        return res === null ? HttpWrapper.Ok({}) : HttpWrapper.FromResult(res);
    }

    static async ValidateInputAsync(record: Promise<HttpWrapper<object>>, result: () => Results | null): Promise<HttpWrapper<object>> {
        return Function.ValidateInput(result);
    }

    static async Init<T>(value: T): Promise<HttpWrapper<T>> {
        return HttpWrapper.Ok(value);
    }

    static async InitAsync<T>(record: Promise<HttpWrapper<T>>): Promise<HttpWrapper<T>> {
        return record;
    }

    static async InitWithRecord<T, R>(record: Promise<HttpWrapper<T>>, value: R): Promise<HttpWrapper<R>> {
        return HttpWrapper.Ok(value);
    }

    static async MapAsync<T, R>(record: Promise<HttpWrapper<T>>, mapper: (value: T) => Promise<HttpWrapper<R>>): Promise<HttpWrapper<R>> {
        const waitedRec = await record;

        if (!waitedRec.continue) {
            return waitedRec.cloneFailed<R>();
        }

        const value = waitedRec.value;

        if (value === null) {
            throw new Error("Mapping null value");
        }

        return await mapper(value);
    }

    static async MapEachAsync<T, R>(record: Promise<HttpWrapper<T[]>>, mapper: (value: T) => Promise<HttpWrapper<R>>): Promise<HttpWrapper<R[]>> {
        const waitedRec = await record;

        if (!waitedRec.continue) {
            return waitedRec.cloneFailed<R[]>();
        }

        const values = waitedRec.value;

        if (values === null) {
            throw new Error("Mapping null value");
        }

        const mappedValues = await Promise.all(values.map(mapper));

        const failed = mappedValues.filter(f => !f.continue).map(s => s.result);
        const success = mappedValues.filter(f => f.continue && f.value !== null).map(s => s.value!);

        return failed.length > 0 ? HttpWrapper.FromResult(failed[0]) : HttpWrapper.Ok(success);
    }

    static async Map<T, R>(record: Promise<HttpWrapper<T>>, mapper: (value: T) => R): Promise<HttpWrapper<R>> {
        const waitedRec = await record;

        if (!waitedRec.continue) {
            return waitedRec.cloneFailed<R>();
        }

        const value = waitedRec.value;

        if (value === null) {
            throw new Error("Mapping null value");
        }

        return HttpWrapper.Ok(mapper(value));
    }

    static async OnResult<T>(record: Promise<HttpWrapper<T>>, result: number, mapper: () => T): Promise<HttpWrapper<T>> {
        const waitedRec = await record;

        if (waitedRec.continue || waitedRec.result.statusCode !== result) {
            return waitedRec;
        }

        return HttpWrapper.Ok(mapper());
    }

    static async OnResultAsync<T>(record: Promise<HttpWrapper<T>>, result: number, mapper: () => Promise<HttpWrapper<T>>): Promise<HttpWrapper<T>> {
        const waitedRec = await record;

        if (waitedRec.continue) {
            return waitedRec;
        }

        return await mapper();
    }
}